File: server/app/models/ops.py
"""内容生产：流程、步骤、运行记录。"""
from datetime import datetime

from app.models.base import Base, Boolean, DateTime, ForeignKey, Integer, Mapped, String, Text, UniqueConstraint, gen_id, mapped_column, utcnow


class OpsStep(Base):
    __tablename__ = "ops_steps"
    id: Mapped[str] = mapped_column(String(32), primary_key=True, default=lambda: gen_id("ost"))
    key: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(128))
    product: Mapped[str] = mapped_column(String(32), index=True)
    step_type: Mapped[str] = mapped_column(String(32), default="script")
    command: Mapped[str] = mapped_column(Text, default="")
    cwd_rel: Mapped[str] = mapped_column(String(128), default="")
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_hint: Mapped[int] = mapped_column(Integer, default=0)
    notes: Mapped[str] = mapped_column(Text, default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)


class OpsPipeline(Base):
    __tablename__ = "ops_pipelines"
    id: Mapped[str] = mapped_column(String(32), primary_key=True, default=lambda: gen_id("opl"))
    key: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(128))
    product: Mapped[str] = mapped_column(String(32), index=True)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    cron: Mapped[str] = mapped_column(String(64), default="")
    timezone: Mapped[str] = mapped_column(String(64), default="Asia/Shanghai")
    notes: Mapped[str] = mapped_column(Text, default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow, onupdate=utcnow)


class OpsPipelineStep(Base):
    __tablename__ = "ops_pipeline_steps"
    __table_args__ = (UniqueConstraint("pipeline_id", "step_id", name="uq_ops_pipeline_step"),)
    id: Mapped[str] = mapped_column(String(32), primary_key=True, default=lambda: gen_id("ops"))
    pipeline_id: Mapped[str] = mapped_column(ForeignKey("ops_pipelines.id"), index=True)
    step_id: Mapped[str] = mapped_column(ForeignKey("ops_steps.id"), index=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)


class OpsRun(Base):
    __tablename__ = "ops_runs"
    id: Mapped[str] = mapped_column(String(32), primary_key=True, default=lambda: gen_id("orn"))
    pipeline_id: Mapped[str] = mapped_column(ForeignKey("ops_pipelines.id"), index=True)
    status: Mapped[str] = mapped_column(String(32), default="running", index=True)
    current_index: Mapped[int] = mapped_column(Integer, default=0)
    log: Mapped[str] = mapped_column(Text, default="")
    started_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class OpsRunStep(Base):
    __tablename__ = "ops_run_steps"
    id: Mapped[str] = mapped_column(String(32), primary_key=True, default=lambda: gen_id("ors"))
    run_id: Mapped[str] = mapped_column(ForeignKey("ops_runs.id"), index=True)
    step_id: Mapped[str] = mapped_column(String(32), default="")
    step_name: Mapped[str] = mapped_column(String(128), default="")
    status: Mapped[str] = mapped_column(String(32), default="pending")
    exit_code: Mapped[int | None] = mapped_column(Integer, nullable=True)
    output: Mapped[str] = mapped_column(Text, default="")
    started_at: Mapped[datetime] = mapped_column(DateTime, default=utcnow)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

File: server/admin-web/src/api/ops.ts
import http, { getData } from './http'

export interface OpsStep {
  id: string
  key: string
  name: string
  product: string
  stepType: string
  command: string
  cwdRel: string
  enabled: boolean
  sortHint: number
  notes: string
  inPipelineEnabled?: boolean
  sortOrder?: number
  linkId?: string
}

export interface OpsPipeline {
  id: string
  key: string
  name: string
  product: string
  enabled: boolean
  cron: string
  timezone: string
  notes: string
  cronNote?: string
  steps: OpsStep[]
}

export interface OpsRun {
  id: string
  pipelineId: string
  pipelineName: string
  status: string
  currentIndex: number
  log: string
  startedAt: string
  finishedAt: string
  materialRoot: string
  steps: { id: string; stepName: string; status: string; exitCode: number | null; output: string }[]
}

export function fetchOpsSteps() {
  return getData<OpsStep[]>(http.get('/admin/ops/steps'))
}

export function patchOpsStep(id: string, data: Partial<OpsStep> & { notes?: string; command?: string }) {
  return getData<OpsStep>(http.patch(`/admin/ops/steps/${id}`, data))
}

export function runOpsStep(id: string) {
  return getData<{ stepId: string; name: string; exitCode: number; output: string; ok: boolean }>(
    http.post(`/admin/ops/steps/${id}/run`),
  )
}

export function fetchOpsPipelines() {
  return getData<OpsPipeline[]>(http.get('/admin/ops/pipelines'))
}

export function patchOpsPipeline(
  id: string,
  data: {
    name?: string
    enabled?: boolean
    cron?: string
    notes?: string
    steps?: { linkId: string; enabled?: boolean; sortOrder?: number }[]
  },
) {
  return getData<OpsPipeline>(http.patch(`/admin/ops/pipelines/${id}`, data))
}

export function runOpsPipeline(id: string) {
  return getData<OpsRun>(http.post(`/admin/ops/pipelines/${id}/run`))
}

export function fetchOpsRuns(pipelineId?: string) {
  return getData<OpsRun[]>(http.get('/admin/ops/runs', { params: pipelineId ? { pipeline_id: pipelineId } : undefined }))
}

export function confirmOpsRun(id: string) {
  return getData<OpsRun>(http.post(`/admin/ops/runs/${id}/confirm`))
}
